mediatek: filogic: add support for ASUS RT-AX59U

xavifr opened the pull request:
(based on support for ASUS RT-AX59U by liushiyou006)

SOC: MediaTek MT7986
RAM: 512MB DDR4
FLASH: 128MB SPI-NAND (Winbond W25N01GV)
WIFI: Mediatek MT7986 DBDC 802.11ax 2.4/5 GHz
ETH: MediaTek MT7531 Switch
UART: 3V3 115200 8N1 (Pinout silkscreened / Do not connect VCC)

Upgrade from AsusWRT to OpenWRT using UART

    Download the OpenWrt initramfs image.
    Copy the image to a TFTP server reachable at 192.168.1.70/24. Rename the image to rtax59u.bin.

    Connect the PC with TFTP server to the RT-AX59U.
    Set a static ip on the ethernet interface of your PC.
    (ip address: 192.168.1.70, subnet mask:255.255.255.0)
    Conect to the serial console, interrupt the autoboot process by pressing '4' when prompted.

    Download & Boot the OpenWrt initramfs image.

    $ setenv ipaddr 192.168.1.1
    $ setenv serverip 192.168.1.70
    $ tftpboot 0x46000000 rtax59u.bin
    $ bootm 0x46000000

    Wait for OpenWrt to boot. Transfer the sysupgrade image to the device using scp and install using sysupgrade.

    $ sysupgrade -n <path-to-sysupgrade.bin>

Upgrade from AsusWRT to OpenWRT using WebUI

    Download transit TRX file from https://drive.google.com/drive/folders/1A20QdjK7Udagu31FSszpWAk8-cGlCwsq

    Upgrade firmware from WebUI (192.168.50.1) using downloaded TRX file

    Wait for OpenWRT to boot (192.168.1.1).

    Upgrade system with sysupgrade image using luci or uploading it through scp and executing sysupgrade command

MAC Address for WLAN 5g is not following the same algorithm as in AsusWRT.
We have increased by one the WLAN 5g to avoid collisions with other networks from WLAN 2g
when bit 28 is already set.

```
                : Stock             : OpenWrt
    WLAN 2g (1) : C8:xx:xx:0D:xx:D4 : C8:xx:xx:0D:xx:D4
    WLAN 2g (2) :                   : CA:xx:xx:0D:xx:D4
    WLAN 2g (3) :                   : CE:xx:xx:0D:xx:D4
    WLAN 5g (1) : CA:xx:xx:1D:xx:D4 : CA:xx:xx:1D:xx:D5
    WLAN 5g (2) :                   : CE:xx:xx:1D:xx:D5
    WLAN 5g (3) :                   : C2:xx:xx:1D:xx:D5
    
    WLAN 2g (1) : 08:xx:xx:76:xx:BE : 08:xx:xx:76:xx:BE
    WLAN 2g (2) :                   : 0A:xx:xx:76:xx:BE
    WLAN 2g (3) :                   : 0E:xx:xx:76:xx:BE
    WLAN 5g (1) : 0A:xx:xx:76:xx:BE : 0A:xx:xx:76:xx:BF
    WLAN 5g (2) :                   : 0E:xx:xx:76:xx:BF
    WLAN 5g (3) :                   : 02:xx:xx:76:xx:BF
```

hauke commented:
Do you plan to support the unmodified U-Boot too?

If you anyway install a modified U-Boot version, it would be nice if you could build it using OpenWrt too.

xavifr commented:
> Do you plan to support the unmodified U-Boot too?

I've added support for the original Asus layout, using liushiyou006 PR and merging with my changes.

> If you anyway install a modified U-Boot version, it would be nice if you could build it using OpenWrt too.

I haven't build that u-boot

tkso1997 commented:
The patch works here without any Problems (unmodified U-Boot)

tkso1997 reviewed (commented):
Anything I can help/test that the device gets accepted?

xavifr commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -11,10 +11,17 @@ mediatek_setup_interfaces()
 	asus,tuf-ax4200)
 		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3 lan4" eth1
 		;;
+<<<<<<< HEAD
```
Oops, sorry!
I've missed some merge hints.

Fixed

xavifr commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -11,10 +11,17 @@ mediatek_setup_interfaces()
 	zyxel,nwa50ax-pro)
 		ucidef_set_interface_lan "eth0"
 		;;
+=======
```
Fixed

xavifr commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -11,10 +11,17 @@ mediatek_setup_interfaces()
 		;;
+=======
+	asus,rt-ax59u|\
+>>>>>>> c47d35a66c (mediatek: filogic: Add support for Asus AX59U with stock layout)
```
Fixed

nuclimedia commented:
Checked locally with stock UBOOT and it works fine

tkso1997 reviewed (approved):
Works fine on my system for the last couple of weeks

hauke commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -105,10 +105,25 @@ define Build/cetron-header
 endef
 
-define Device/asus_tuf-ax4200
+define Device/acer_predator-w6
```
Please add a seperate commit before this one which moves the `acer_predator-w6` device to the position you moved it to to make it easier to review your changes.

xavifr replied:
done!

hauke commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/firmware/11-mt76-caldata:
```diff
@@ -45,10 +45,15 @@ case "$FIRMWARE" in
-	asus,tuf-ax4200)
+	asus,rt-ax59u|\
+	asus,tuf-ax4200|\
+	asus,tuf-ax6000)
```
Did you intentionally add the `asus,tuf-ax6000` here?

xavifr replied:
It was an copy&paste error, sorry

hauke commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,271 @@
+		};
+
+		phy6: phy@6 {
+			compatible = "ethernet-phy-id67c9.de0a";
```
I think this PHY is not used. I do not see any references to it. If it is not used please remove this entry.

xavifr replied:
It was an copy&paste error, sorry

hauke commented:
Please add to the commit message what was modified in the modified u-boot and where to get it from. 

Please also extend the installation description, maybe take some parts from this description:
https://github.com/openwrt/openwrt/pull/12894/commits/a413b0618d27d8863cdbbdb977862a24974643dd

If the installation is different when the stock u-boot and the modified one is used please describe both ways.

xavifr committed:
mediatek: filogic: add support ASUS RT-AX59U

(based on support for ASUS RT-AX59U by liushiyou006)

SOC: MediaTek MT7986
RAM: 512MB DDR4
FLASH: 128MB SPI-NAND (Winbond W25N01GV)
WIFI: Mediatek MT7986 DBDC 802.11ax 2.4/5 GHz
ETH: MediaTek MT7531 Switch
UART: 3V3 115200 8N1 (Pinout silkscreened / Do not connect VCC)

Upgrade from AsusWRT to OpenWRT using UART

    Download the OpenWrt initramfs image.
    Copy the image to a TFTP server reachable at 192.168.1.70/24. Rename the image to rtax59u.bin.

    Connect the PC with TFTP server to the RT-AX59U.
    Set a static ip on the ethernet interface of your PC.
    (ip address: 192.168.1.70, subnet mask:255.255.255.0)
    Conect to the serial console, interrupt the autoboot process by pressing '4' when prompted.

    Download & Boot the OpenWrt initramfs image.

    $ setenv ipaddr 192.168.1.1
    $ setenv serverip 192.168.1.70
    $ tftpboot 0x46000000 rtax59u.bin
    $ bootm 0x46000000

    Wait for OpenWrt to boot. Transfer the sysupgrade image to the device using scp and install using sysupgrade.

    $ sysupgrade -n <path-to-sysupgrade.bin>

Upgrade from AsusWRT to OpenWRT using WebUI

    Download transit TRX file from https://drive.google.com/drive/folders/1A20QdjK7Udagu31FSszpWAk8-cGlCwsq

    Upgrade firmware from WebUI (192.168.50.1) using downloaded TRX file

    Wait for OpenWRT to boot (192.168.1.1).

    Upgrade system with sysupgrade image using luci or uploading it through scp and executing sysupgrade command

MAC Address for WLAN 5g is not following the same algorithm as in AsusWRT.
We have increased by one the WLAN 5g to avoid collisions with other networks from WLAN 2g
when bit 28 is already set.

              : Stock             : OpenWrt
  WLAN 2g (1) : C8:xx:xx:0D:xx:D4 : C8:xx:xx:0D:xx:D4
  WLAN 2g (2) :                   : CA:xx:xx:0D:xx:D4
  WLAN 2g (3) :                   : CE:xx:xx:0D:xx:D4
  WLAN 5g (1) : CA:xx:xx:1D:xx:D4 : CA:xx:xx:1D:xx:D5
  WLAN 5g (2) :                   : CE:xx:xx:1D:xx:D5
  WLAN 5g (3) :                   : C2:xx:xx:1D:xx:D5

  WLAN 2g (1) : 08:xx:xx:76:xx:BE : 08:xx:xx:76:xx:BE
  WLAN 2g (2) :                   : 0A:xx:xx:76:xx:BE
  WLAN 2g (3) :                   : 0E:xx:xx:76:xx:BE
  WLAN 5g (1) : 0A:xx:xx:76:xx:BE : 0A:xx:xx:76:xx:BF
  WLAN 5g (2) :                   : 0E:xx:xx:76:xx:BF
  WLAN 5g (3) :                   : 02:xx:xx:76:xx:BF

Signed-off-by: Xavier Franquet <xavier@franquet.es>

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,271 @@
+
+		phy5: phy@5 {
+			compatible = "ethernet-phy-id67c9.de0a";
+			reg = <5>;
```
removed

dmemarko commented:
"Simple" installation is also possible, by the way: https://forum.openwrt.org/t/openwrt-support-for-asus-rt-ax59u/154666/13

Wiedzmin89 commented:
> "Simple" installation is also possible, by the way: https://forum.openwrt.org/t/openwrt-support-for-asus-rt-ax59u/154666/13

This method is based on [fork](https://github.com/openwrt-xiaomi/openwrt), where you can build *-initramfs.trx images.
It also has its own ASUS RT-AX59U implementation: https://github.com/openwrt-xiaomi/openwrt/commit/42d7c3621662ebfc7e4193a88a0bd7dff5c2d82d

remittor commented:
@xavifr , Please expand the available space by removing the `jffs2` volume: https://github.com/openwrt-xiaomi/openwrt/commit/07f79132c2181298eece9d589670293066ef33ce

xavifr commented:
> @xavifr , Please expand the available space by removing the `jffs2` volume: [openwrt-xiaomi@07f7913](https://github.com/openwrt-xiaomi/openwrt/commit/07f79132c2181298eece9d589670293066ef33ce)

done! not it has 106mb free space at the end
![Screenshot from 2023-11-22 15-35-49](https://github.com/openwrt/openwrt/assets/13116609/820b4b3f-c81e-40a1-b19d-63e3d0d3ff42)

xavifr commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -120,6 +120,43 @@ define Device/acer_predator-w6
+  DEVICE_MODEL := RT-AX59U (Alternative U-Boot layout)
+  DEVICE_DTS := mt7986a-asus-rt-ax59u-ubootmod
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_DTS_LOADADDR := 0x47000000
```
It boots OK without that line, I've also removed it from RT-AX59U-stock

xavifr commented:
> "Simple" installation is also possible, by the way: https://forum.openwrt.org/t/openwrt-support-for-asus-rt-ax59u/154666/13

Tested with https://drive.google.com/drive/folders/1A20QdjK7Udagu31FSszpWAk8-cGlCwsq and stock version.

It is also possible to upload the initial image without entering the webui.
Procedure:
- connect cable to lanX port with ip in range 192.168.1.0/24 (avoid 192.168.1.1)
- keep the reset button pressed
- power on router, led will blink purple
- ping to 192.168.1.1, should work
- execute `atftp --tftp-timeout 1 -g -r ASUSSPACELINK$(echo -ne "\x01\x01\xa8\xc0") -l /dev/null 192.168.1.1`
- execute `atftp -p -l <TRANSIT_IMAGE> -r ASUSSPACELINK 192.168.1.1` where <TRANSIT_IMAGE> is the TRX file downloaded from google drive
- router should boot with that transit image, then you can put any other final image

csharper2005 commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-stock.dts:
```diff
@@ -0,0 +1,22 @@
+};
+
+&partitions {
+    partition@0 {
+        label = "Bootloader";
+        reg = <0x0 0x400000>;
+        read-only;
+    };
+
+    partition@400000 {
+        label = "UBI_DEV";
+        reg = <0x400000 0x7c00000>;
+    };
+};
```
Please fix wrong indents.

csharper2005 reviewed (commented):
Thanks for your work! I'm also interesting in this device.

Your commit doesn't contain MAC addresses table. Are all MAC addresses in OpenWrt identical to the stock ones?

Do I understand correctly that the installation procedure is the same for stock and alternative layout?

xavifr commented:
> Thanks for your work! I'm also interesting in this device.
> 
> Your commit doesn't contain MAC addresses table. Are all MAC addresses in OpenWrt identical to the stock ones?
> 
> Do I understand correctly that the installation procedure is the same for stock and alternative layout?

Thanks for the comments, I will check them on Monday.

Mac addresses are the same as in asuswrt.

For the installation procedure, do not use ubootmod if you have a stock device.

csharper2005 commented:
> Thanks for the comments, I will check them on Monday.
> Mac addresses are the same as in asuswrt.

Thanks!

> For the installation procedure, do not use ubootmod if you have a stock device.

I've never seen before merged patches where device require `ubootmod` and `ubootmod` is not included in the merge request. @hauke, @dangowrt what is the current bootloader replacement policy for this target?

csharper2005 commented:
I managed install the OpenWrt without UART (using direct initramfs writing from the stock firmware). Can anyone explain me, please, why do we need alternate U-Boot? 

Free space for the packages:
![image](https://github.com/openwrt/openwrt/assets/82600916/ba453a6d-a777-4f8a-bb89-3e6693363df0)

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-ubootmod.dts:
```diff
@@ -0,0 +1,42 @@
+	partition@380000 {
+		label = "FIP";
+		reg = <0x380000 0x0200000>;
+	};
```
done

xavifr commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -105,6 +105,56 @@ define Build/cetron-header
+  KERNEL := kernel-bin | lzma | \
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  KERNEL_INITRAMFS := kernel-bin | lzma | \
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 64k
```
done

xavifr commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -105,6 +105,56 @@ define Build/cetron-header
+  DEVICE_DTS := mt7986a-asus-rt-ax59u-ubootmod
+  DEVICE_DTS_DIR := ../dts
+  DEVICE_PACKAGES := kmod-usb3 kmod-mt7986-firmware mt7986-wo-firmware
+  IMAGES := sysupgrade.bin
```
done

xavifr commented on target/linux/mediatek/image/filogic.mk:
```diff
@@ -105,6 +105,56 @@ define Build/cetron-header
+  KERNEL := kernel-bin | lzma | \
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb
+  KERNEL_INITRAMFS := kernel-bin | lzma | \
+	fit lzma $$(KDIR)/image-$$(firstword $$(DEVICE_DTS)).dtb with-initrd | pad-to 64k
```
done

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,257 @@
+		led_blue: blue {
+			label = "blue:system";
+			gpios = <&pio 13 GPIO_ACTIVE_LOW>;
+		};
```
changed

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,257 @@
+
+		mesh {
+			label = "wps";
+			gpios = <&pio 9 GPIO_ACTIVE_LOW>;
```
done

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,257 @@
+	keys {
+		compatible = "gpio-keys";
+
+		reset {
```
done

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-ubootmod.dts:
```diff
@@ -0,0 +1,42 @@
+		label = "UBI_DEV";
+		reg = <0x580000 0x4000000>;
+	};
+
```
done

xavifr commented:
> I managed install the OpenWrt without UART (using direct initramfs writing from the stock firmware). Can anyone explain me, please, why do we need alternate U-Boot?
> 
> Free space for the packages: ![image](https://private-user-images.githubusercontent.com/82600916/285966187-ba453a6d-a777-4f8a-bb89-3e6693363df0.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTEiLCJleHAiOjE3MDExNTU3NzIsIm5iZiI6MTcwMTE1NTQ3MiwicGF0aCI6Ii84MjYwMDkxNi8yODU5NjYxODctYmE0NTNhNmQtYTc3Ny00ZjhhLWJiODktM2U2NjkzMzYzZGYwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFJV05KWUFYNENTVkVINTNBJTJGMjAyMzExMjglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjMxMTI4VDA3MTExMlomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPTIwYTBjNjg0YjU1OGJiYjY3YjdiNjJkNTliNjQxYTIzYmI2MGRlMGFlYjkwYjc5ODZmMWYwMGFkMGMzMDc2ZjImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.IF8nuzue3YqPsZqv6XI_3v23dp5wNsWzP-F6Rmn3-Ik)

After the appearance of this new method with the TRX, the ubootmod version may no longer be necessary.

remittor commented:
> new method with the TRX

But that’s not the reason.
It’s just that when adding device `TUF-AX4200`, the maintainer @blocktrron had to look at how device `Redmi AX6000` was added to the OpenWRT repository.

https://github.com/openwrt-xiaomi/openwrt/blob/a33b705e225c53166c22c6201552b597d61a43f5/target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh#L177-L190

xavifr commented:
> something is wrong with the generation of /etc/board.json. Ports are listed twice including a "lan4" which is non-existend on this device. Not sure if it is a current OpenWrt issue or it has something todo with the patch set.


I've tested it right now, applying the changes into 23.05 and the board.json worked as expected
![image](https://github.com/openwrt/openwrt/assets/13116609/f3a3ea11-80ba-4587-a6f0-87393b9d5344)

It also works with main branch

![image](https://github.com/openwrt/openwrt/assets/13116609/94932683-6c49-45df-8498-62f670857219)

csharper2005 commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,261 @@
+			color = <LED_COLOR_ID_GREEN>;
+			function = LED_FUNCTION_STATUS;
+			gpios = <&pio 13 GPIO_ACTIVE_LOW>;
+		};
```
Have you already tested the leds? They don't work on my device.

csharper2005 replied:
@xavifr sorry, I suggested you the code with mistake. Should be `color = <LED_COLOR_ID_BLUE>;` for the blue led. That's the reason.

csharper2005 replied:
Upd. Green led is blinking with initramfs and don't work after writing to flash. That's very strange... I have no idea why yet.

xavifr replied:
ok

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -81,13 +85,20 @@ mediatek_setup_macs()
 		wan_mac="${addr}"
 		lan_mac="${addr}"
 		;;
+	asus,rt-ax59u-ubootmod)
+		CI_UBIPART="UBI_DEV"
+		addr=$(mtd_get_mac_binary "Factory" 0x4)
+		wan_mac="${addr}"
+		lan_mac="${addr}"
+		;;
```
Do you really need this? You have `macaddr_factory_4` in dts and can assign the lan and wan mac also in dts.

csharper2005 replied:
I think it's the easiest and and at the same time the best choice.

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -16,6 +16,7 @@ case "$board" in
 		[ "$PHYNBR" = "1" ] && cat $key_path/6gMAC > /sys${DEVPATH}/macaddress
 		[ "$PHYNBR" = "2" ] && cat $key_path/5gMAC > /sys${DEVPATH}/macaddress
 		;;
+	asus,rt-ax59u|\
```
I got wrong MACs with this code.
```
                :  Stock            : OpenWrt (this PR)
MAC address 2g	: 08:xx:xx:0D:xx:FE : 0A:xx:xx:0D:xx:FF
MAC address 5g	: 0A:xx:xx:1D:xx:FE : 08:xx:xx:0D:xx:FE
```

The working one for my device:
```
	asus,rt-ax59u)
		CI_UBIPART="UBI_DEV"
		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
		[ "$PHYNBR" = "1" ] && \
			macaddr_setbit_la $(macaddr_setbit $addr 7) > /sys${DEVPATH}/macaddress
		;;
```

xavifr replied:
I've tested another AX59U with original Asus firmwares and that is my MAC table:
```
                 :  Stock 3.0.0.4.288 31345 and 32702
MAC address 2g   : C8:7F:XX:XX:XX:BE
MAC address 5g   : CA:7F:XX:XX:XX:BE
MAC address LAN2 : C8:7F:XX:XX:XX:BE
MAC address WAN  : C8:7F:XX:XX:XX:BE
```

Maybe we should only use setbit_la in phy1

```
        asus,rt-ax59u)
                CI_UBIPART="UBI_DEV"
                addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
                [ "$PHYNBR" = "1" ] && \
                        macaddr_setbit_la $addr > /sys${DEVPATH}/macaddress
                ;;

```

csharper2005 replied:
> Maybe we should only use `setbit_la` in phy1

It's possible. The flaw is that the 2nd(3rd, ...) 2.4 SSID will be equal to the 1st(2nd, ...) 5 GHz SSID

> : Stock 3.0.0.4.288 31345 and 32702
> MAC address 2g   : C8:7F:XX:XX:XX:BE
> MAC address 5g   : CA:7F:XX:XX:XX:BE
> MAC address LAN2 : C8:7F:XX:XX:XX:BE
> MAC address WAN  : C8:7F:XX:XX:XX:BE

Could you show the 4th byte of the MAC, please? I also checked both 3.0.0.4.388_31345 and 3.0.0.4.388_32702.

```
                                :  Stock
MAC address 2g	: 08:xx:xx:0D:xx:FE
MAC address 5g	: 0A:xx:xx:1D:xx:FE
LAN             : 08:xx:xx:0D:xx:FE
WAN             : 08:xx:xx:0D:xx:FE
```

We need more statistics. @tkso1997 can you share your MACs, please?

xavifr replied:
```
                 : Stock 3.0.0.4.288 31345 and 32702
MAC address 2g   : C8:7F:XX:76:XX:BE
MAC address 5g   : CA:7F:XX:76:XX:BE
MAC address LAN2 : C8:7F:XX:76:XX:BE
MAC address WAN  : C8:7F:XX:76:XX:BE
```

I have more routers here, I'm going to check some more routers and will post results here.

csharper2005 replied:
@xavifr, thanks. Mow everything is clear. C8:7F:XX:76:XX:BE already has 1 in 28th bit. As a result the 4th bytes are identical.

My proposal is:
1. Keep original ASUS algorithm (and set 28th bit);
2. Increase by one the 5g MAC to avoid 2g and 5g MAC addresses matching (this should be documented in the commit message). You 5g MAC will be `C8:7F:XX:76:XX:BF`.
```
	asus,rt-ax59u)
		CI_UBIPART="UBI_DEV"
		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
		[ "$PHYNBR" = "1" ] && \
			macaddr_setbit_la $(macaddr_setbit $(macaddr_add $addr 1) 28) > \
			/sys${DEVPATH}/macaddress
		;;

```

xavifr replied:
I can confirm your thoughts, 

```
                 : Stock 3.0.0.4.288 31345
MAC address 2g   : 08:BF:XX:0D:XX:56
MAC address 5g   : 0A:BF:XX:1D:XX:56
MAC address LAN2 : 08:BF:XX:0D:XX:56
MAC address WAN  : 08:BF:XX:0D:XX:56
```

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/hotplug.d/ieee80211/11_fix_wifi_mac:
```diff
@@ -26,6 +27,11 @@ case "$board" in
 		[ "$PHYNBR" = "1" ] && echo "$addr" > /sys${DEVPATH}/macaddress
 		;;
+	asus,rt-ax59u-ubootmod)
+		addr=$(mtd_get_mac_binary "Factory" 0x4)
+		[ "$PHYNBR" = "0" ] && macaddr_setbit_la $(macaddr_add $addr 1) > /sys${DEVPATH}/macaddress
+		[ "$PHYNBR" = "1" ] && echo "$addr" > /sys${DEVPATH}/macaddress
+		;;
```
The same as above. MAC address table in the commit message will dispel doubts.

xavifr replied:
done

csharper2005 commented on target/linux/mediatek/filogic/base-files/lib/preinit/10_fix_eth_mac.sh:
```diff
@@ -10,6 +10,15 @@ preinit_set_mac_address() {
+		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
+		ip link set dev eth0 address "$addr"
+		;;
+	asus,rt-ax59u-ubootmod)
+		addr=$(mtd_get_mac_binary "Factory" 0x4)
+		ip link set dev eth0 address "$addr"
+		;;
```
You can define this MAC via dts for the ubootmod version.

csharper2005 reviewed (changes requested):
So, the main topics:
- MACs
- LEDs (if some functionality doesn't work then please describe this in the commit message)
- Installation guide
- PR requires rebase due to merge conflict

P.S. Typo in the 1st commit title:
mediatek: filogi**n**: move target acer_predator before asus_tuf

csharper2005 commented:
> After the appearance of this new method with the TRX, the ubootmod version may no longer be necessary.

If I understand correctly the TRX method is intended to simplify installation. We need @remittor's PR to the firmware tools merged  to add trx image in future.

We have already got whole free space using UART installation (or initramfs direct writing) with stock u-boot. Nothing more is required. 👌

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-stock.dts:
```diff
@@ -0,0 +1,22 @@
+#include "mt7986a-asus-rt-ax59u.dtsi"
+
+/ {
+	model = "ASUS RT-AX59U (Stock U-Boot layout)";
```
ok

blocktrron reviewed (changes requested):
Please remove the U-Boot modification from the commit of the original support. You can add it to a seperate commit though and document installation and how to build U-Boot there.

csharper2005 commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-ubootmod.dts:
```diff
@@ -0,0 +1,59 @@
+};
+
+&wifi {
+	mediatek,mtd-eeprom = <&factory 0x0>;
```
New style would be nice:
```suggestion
	nvmem-cells = <eeprom_factory>;
	nvmem-cell-names = "eeprom";
```

csharper2005 commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-ubootmod.dts:
```diff
@@ -0,0 +1,59 @@
+		reg = <0x0100000 0x0080000>;
+	};
+
+	factory: partition@180000 {
```
We can drop this label:
```suggestion
	partition@180000 {
```

csharper2005 commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,272 @@
+	keys {
+		compatible = "gpio-keys";
+
+		button_wps: button-0 {
```
We can drop unreferenced label.
```suggestion
		button-0 {
```

csharper2005 commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,272 @@
+			linux,code = <KEY_WPS_BUTTON>;
+		};
+
+		button_reset: button-1 {
```
We can drop unreferenced label:
```suggestion
		button-1 {
```

xavifr replied:
ok

csharper2005 reviewed (changes requested):
Some suggestions since https://github.com/openwrt/openwrt/pull/13837 is merged.

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,272 @@
+		};
+
+		led_status_blue: led-2 {
+			label = "blue:status";
```
ok

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dtsi:
```diff
@@ -0,0 +1,272 @@
+		};
+
+		led_status_red: led-1 {
+			label = "red:status";
```
ok

xavifr commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u-ubootmod.dts:
```diff
@@ -0,0 +1,42 @@
+	factory: partition@180000 {
+		label = "Factory";
+		reg = <0x180000 0x0200000>;
+	};
```
I've removed ubootmod from the PR

xavifr commented:
> Please remove the U-Boot modification from the commit of the original support. You can add it to a seperate commit though and document installation and how to build U-Boot there.

I've removed the `ubootmod` target from the PR

DragonBluep commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dts:
```diff
@@ -0,0 +1,285 @@
+		switch@0 {
+			compatible = "mediatek,mt7531";
+			reg = <31>;
+			reset-gpios = <&pio 5 0>;
```
Replace `0` with `GPIO_ACTIVE_HIGH`. And are you sure it is not `1`, aka `GPIO_ACTIVE_LOW`

csharper2005 replied:
In original ASUS dts:
```
			switch@0 {
				compatible = "mediatek,mt7531";
				reg = <0x1f>;
				reset-gpios = <0x12 0x05 0x00>;
```
What can be affected by incorrect HIGH/LOW here?

DragonBluep replied:
> What can be affected by incorrect HIGH/LOW here?

The switch won't work at all.

DragonBluep replied:
--to @dangowrt Hi!

I found that mt7531 switch reset code has some issues. I think it should set 1 (assert) first and then set 0 (de-assert). Meanwhile, we should use `GPIO_ACTIVE_LOW` instead of `GPIO_ACTIVE_HIGH` in the dts. And some delay after reset should be necessary. I saw @ptpt52 make a such hack in his repo https://github.com/x-wrt/x-wrt/commit/49570fdf9b0114840d0c70021fe2e6ec561241f0.

Looking forward to your reply.

Reported-by: Shiji Yang <yangshiji66@outlook.com>

Ref:
mt7531 reset:
https://github.com/torvalds/linux/blob/bee0e7762ad2c6025b9f5245c040fcc36ef2bde8/drivers/net/dsa/mt7530.c#L2249-L2251
```
		gpiod_set_value_cansleep(priv->reset, 0);
		usleep_range(1000, 1100);
		gpiod_set_value_cansleep(priv->reset, 1);
```

qca8k reset:
https://github.com/torvalds/linux/blob/b85ea95d086471afb4ad062012a4d73cd328fa86/drivers/net/dsa/qca/qca8k-8xxx.c#L2040-L2052
```
		gpiod_set_value_cansleep(priv->reset_gpio, 1);
		/* The active low duration must be greater than 10 ms
		 * and checkpatch.pl wants 20 ms.
		 */
		msleep(20);
		gpiod_set_value_cansleep(priv->reset_gpio, 0);
```

csharper2005 replied:
> > What can be affected by incorrect HIGH/LOW here?
> 
> The switch won't work at all.

The switch is working with `GPIO_ACTIVE_HIGH`

xavifr replied:
I've also tested with `GPIO_ACTIVE_HIGH` and switch is working OK.

I've changed it in code

DragonBluep commented on target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh:
```diff
@@ -1,5 +1,16 @@
+	ubirmvol /dev/ubi0 -N rootfs
+	ubirmvol /dev/ubi0 -N rootfs_data
+	ubirmvol /dev/ubi0 -N jffs2
+	ubimkvol /dev/ubi0 -N jffs2 -s 0x3e000
```
Why create a jffs2 volume?

DragonBluep replied:
Oh, I got it. I would suggest check some partition label / size or magic number instead of the `$(rootfs_type)` to identify whether it is an official image. If my understanding is correct each time running initramfs will format the user configs. This is unacceptable because someone like to use initramfs to test new features.

csharper2005 replied:
or just add the commands to the installation guide...

remittor replied:
> If my understanding is correct each time **running initramfs** will format the user configs. This is unacceptable because someone like to use initramfs to test new features.

And you search through the entire repository and understand that this is exactly how it is done everywhere.
1) https://github.com/openwrt/openwrt/blob/62277c485e79f6fe453a8a1bd54f271ae17c7490/target/linux/mediatek/filogic/base-files/lib/upgrade/platform.sh#L20
2) https://github.com/openwrt/openwrt/blob/62277c485e79f6fe453a8a1bd54f271ae17c7490/target/linux/qualcommax/ipq807x/base-files/lib/upgrade/buffalo.sh#L22
3) https://github.com/openwrt/openwrt/blob/62277c485e79f6fe453a8a1bd54f271ae17c7490/target/linux/ipq806x/base-files/lib/upgrade/buffalo.sh#L42
4) etc...

You can launch the `initramfs` image in two different ways:
1) through special bootloader commands, copying the image directly into memory
2) use `sysupgrade` to write the image to the flash chip

The `asus_initial_setup` function will only be written in the second case.

patrykk replied:
> Oh, I got it. I would suggest check some partition label / size or magic number instead of the `$(rootfs_type)` to identify whether it is an official image. If my understanding is correct each time running initramfs will format the user configs. This is unacceptable because someone like to use initramfs to test new features.

@DragonBluep  I fully agree with you. 
@remittor We need to implement it in other manner for example as a separate package or script which can be executed manually then tester will not lose data.

remittor replied:
@patrykk , @DragonBluep 
Why is partition `overlay` boldly deleted on all Xiaomi devices without asking questions, but for Asus routers you need to ask the user?

DragonBluep replied:
There is no special reason. Just because I didn't review the Xiaomi PR. I'm not a repo maintainer. And it's fine for me if only case 3 trigger the script.

> use sysupgrade to write the tar image to the flash chip (on initramfs mode)

csharper2005 commented on target/linux/mediatek/filogic/base-files/etc/board.d/02_network:
```diff
@@ -11,6 +11,9 @@ mediatek_setup_interfaces()
 		;;
+	asus,rt-ax59u)
+		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3" wan
+		;;
```
Can be merged with this existing block:
```
	cetron,ct3003|\
	confiabits,mt7981|\
	cudy,wr3000-v1|\
	jcg,q30-pro|\
	qihoo,360t7)
		ucidef_set_interfaces_lan_wan "lan1 lan2 lan3" wan
		;;
```

csharper2005 replied:
@xavifr thanks for your updates. 

A few words about the commit message：
- Please add UART installation method too because trx file at the drive.google.com can become unreachable in future. And the commit message is a primary source of the guide if we have no wiki page.
- Please add note about MAC addresses and differences from the original Asus algorithm.

xavifr replied:
done

ptpt52 commented:
ping @981213

csharper2005 reviewed (approved):
LGTM now.

csharper2005 commented on target/linux/mediatek/filogic/base-files/lib/preinit/10_fix_eth_mac.sh:
```diff
@@ -10,6 +10,11 @@ preinit_set_mac_address() {
 		ip link set dev game address "$(cat $key_path/LANMAC)"
 		ip link set dev eth1 address "$(cat $key_path/WANMAC)"
 		;;
+	asus,rt-ax59u)
+		CI_UBIPART="UBI_DEV"
+		addr=$(mtd_get_mac_binary_ubi "Factory" 0x4)
+		ip link set dev eth0 address "$addr"
+		;;
```
We still have random mac on `eth0`. This don't work for the Mediatek switch. I noticed this on several devices. So we can drop this block because it's useless. We set correct MAC in `02_network` script.

P.S. `eth0` string in `asus,tuf-*` block is also useless.

remittor commented:
Is this fix needed here or not?
https://github.com/openwrt/openwrt/commit/b95e9dae43dfc7b6d3c1b4a031a23643a96246f6

xavifr commented:
> Is this fix needed here or not? [b95e9da](https://github.com/openwrt/openwrt/commit/b95e9dae43dfc7b6d3c1b4a031a23643a96246f6)

already applied

csharper2005 commented:
Rebase with the main branch is required due to the merge conflict.

csharper2005 commented:
@xavifr are you planning to do a backport to the 23.05 branch?

xavifr commented:
> @xavifr are you planning to do a backport to the 23.05 branch?

I've opened a new pull request with backport to 23.05
https://github.com/openwrt/openwrt/pull/14327

csharper2005 commented:
> > @xavifr are you planning to do a backport to the 23.05 branch?
> 
> I've opened a new pull request with backport to 23.05 #14327

Great. Thanks.

rmilecki commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dts:
```diff
@@ -0,0 +1,263 @@
+		bootargs-override = "ubi.mtd=UBI_DEV";
+	};
+
+	memory {
```
This node is missing address (`@40000000`) and `device_type`

melroy89 replied:
owow. Should we create a new PR? Since this is merged already.

rmilecki commented on target/linux/mediatek/dts/mt7986a-asus-rt-ax59u.dts:
```diff
@@ -0,0 +1,263 @@
+
+		spi-max-frequency = <20000000>;
+		spi-tx-buswidth = <4>;
+		spi-rx-buswidth = <4>;
```
Those properties are undocumented and unsupported. Check `Documentation/devicetree/bindings/spi/spi-peripheral-props.yaml` for proper properties.
